### PR TITLE
Enumerables

### PR DESCRIPTION
**Pull request by haraphat01**
This is my final project on Enumerables, this project was built with only with Ruby. The following methods were created.

- My_each

- my_each_with_index

- my_select

- my_all?

- my_any?

- my_none?

- my_count and others

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -1,41 +1,106 @@
...
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    array = to_a
+    while i < array.length
+      yield(array[i])
+      i += 1
     end
     array
   end
```
Comment (Dandush03) [inline enumerables.rb]: ### #my_each

- [ ] When a block is given when self is a range returns the range itself after calling the given block once for each element in self

File: enumerables.rb [hunk 2]
```diff
@@ -1,41 +1,106 @@
...
+
     i = 0
     array = to_a
     while i < array.length
       yield(array[i], i)
-      i+=1
+      i += 1
+    end
+    array
   end
```
Comment (Dandush03) [inline enumerables.rb]: ### #my_each_with_index

- [ ] When a block is given when self is a range returns the range itself after calling the given block once for each element in self

File: enumerables.rb [hunk 3]
```diff
@@ -1,41 +1,106 @@
...
 
-puts [1,2,3,4,5,6].my_select(&:even?)
+  def my_all?
+    return to_enum(:my_all?) unless block_given?
 
+    my_each do |x|
+      return false if yield(x) != true
+    end
+    true
+  end
```
Comment (Dandush03) [inline enumerables.rb]: ### #my_all?
- [ ] When no block or argument is given returns true when none of the collection members are false or nil
- [ ] When no block or argument is given returns false when one of the collection members is false or nil
- [ ] When a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] When a class is passed as an argument returns false if any of the collection is not a member of such class
- [ ] When a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [ ] When a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

File: enumerables.rb [hunk 4]
```diff
@@ -1,41 +1,106 @@
...
+
+  def my_any?
+    return to_enum(:my_any?) unless block_given?
+
+    my_each do |x|
+      return true if yield(x) == true
+    end
+    false
+  end
```
Comment (Dandush03) [inline enumerables.rb]: ### #my_any?
- [ ] When no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] When no block or argument is given returns false if at least one of the collection is not true
- [ ] When a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] When a Regex is passed as an argument returns true if any of the collection matches the Regex
- [ ] When a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [ ] When a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

File: enumerables.rb [hunk 5]
```diff
@@ -1,41 +1,106 @@
...
+
+  def my_none?
+    return to_enum(:my_none?) unless block_given?
+
+    my_each do |x|
+      return false if yield(x) == true
+    end
+    true
+  end
```
Comment (Dandush03) [inline enumerables.rb]: ### #my_none?
- [ ] When no block or argument is given returns true only if none of the collection members is true
- [ ] When no block or argument is given returns false only if one of the collection members is true
- [ ] When a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] When a class is passed as an argument returns false if any of the collection is a member of such class
- [ ] When a Regex is passed as an argument returns true if none of the collection matches the Regex
- [ ] When a Regex is passed as an argument returns false if any of the collection matches the Regex
- [ ] When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [ ] When a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: enumerables.rb [hunk 6]
```diff
@@ -1,41 +1,106 @@
...
+    return to_enum(:my_map) unless block_given?
+
+    map = []
+    if !proc_0.nil?
+      my_each { |value| map.push(proc_0.call(value)) }
+    elsif block_given?
+      my_each { |value| map.push(yield(value)) }
+    end
+    map
+  end
```
Comment (Dandush03) [inline enumerables.rb]: ### #my_map
- [ ] It executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

**Review by Dandush03 — COMMENTED** [review]
# Status: 🔴 Required Changes 🔴 
## Hey!! 👋 

### Great Job on this first attempt

There're some method that are behaving different from what it is expected, please make sure to fix the commented issues below before asking for another review

## Happy Coding! 💻

**Comment by haraphat01** [comment]
Thank you for the review, here's my final project on enumerables for you kind approval.

I've reworked on my_each, my_each_index, and my_map.

I also checked the My_all, my_any, and my_none, they all seem to be working fine here.

File: enumerables.rb [hunk 7]
```diff
@@ -1,41 +1,103 @@
...
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    array = to_a
+    until i >= array.length
+      yield(array[i])
+      i += 1
     end
     array
   end
```
Comment (Torres-ssf) [inline enumerables.rb]: - [ ] my_each when a block is given when self is a range after calling the given block once for each element in self, returns the range itself. Your method is returning an array, even when calling my_each from a range

Example:
`(1..3).my_each { | value | puts value }` should print 1, 2 and 3, and after that should return the range itself `1..3`

Print from the original method:

Screenshot from 2020-06-05 09-19-36

File: enumerables.rb [hunk 8]
```diff
@@ -1,41 +1,103 @@
...
     i = 0
     array = to_a
-    while i < array.length
+    until i >= array.length
       yield(array[i], i)
-      i+=1
+      i += 1
+    end
+    array
   end
```
Comment (Torres-ssf) [inline enumerables.rb]: - [ ] my_each_with_index when a block is given when self is a range after calling the given block once for each element in self, returns the range itself. Your method is returning an array, even when calling my_each_with_index from a range

Example:
`(1..3).my_each_with_index { | value, index | puts index }` should print 0, 1 and 2, and after that should return the range itself `1..3`

Print from the original method:

Screenshot from 2020-06-05 09-22-33

File: enumerables.rb [hunk 9]
```diff
@@ -1,41 +1,103 @@
...
 
-puts [1,2,3,4,5,6].my_select(&:even?)
+  def my_all?
+    return to_enum(:my_all?) unless block_given?
+
+    my_each do |x|
+      return false if yield(x) != true
+    end
+    true
+  end
```
Comment (Torres-ssf) [inline enumerables.rb]: - [ ] Only `each`, `each_with_index`, `select`, and `map` should return `to_enum` when a block is not passed. `all?`, `any?` `none?`, they all do verifications when they are called from an enumerable without a block.

**Use this page as a [guide](https://apidock.com/ruby/Enumerable/all%3F)**

- [ ] When no block or argument is given returns true when none of the collection members are false or nil

**expected output:**
Screenshot from 2020-06-10 10-27-06


- [ ] When no block or argument is given returns false when one of the collection members is

**expected output:**
Screenshot from 2020-06-10 10-29-37

- [ ] When a class is passed as an argument returns true if all of the collection is a member of such class

**expected output:**
Screenshot from 2020-06-10 10-31-32

- [ ] When a class is passed as an argument returns false if any of the collection is not a member of such class

**expected output:**
Screenshot from 2020-06-10 10-35-13


- [ ] When a Regex is passed as an argument returns true if all of the collection matches the Regex

**expected output:**
Screenshot from 2020-06-10 10-42-36

- [ ] When a Regex is passed as an argument returns false if any of the collection does not match the Regex

**expected output:**
Screenshot from 2020-06-10 10-46-12

- [ ] When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

**expected output:**
Screenshot from 2020-06-10 10-50-49

- [ ] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

**expected output:**
Screenshot from 2020-06-10 10-51-57

File: enumerables.rb [hunk 10]
```diff
@@ -1,41 +1,103 @@
...
 
+  def my_any?
+    return to_enum(:my_any?) unless block_given?
 
+    my_each do |x|
+      return true if yield(x) == true
+    end
+    false
+  end
```
Comment (Torres-ssf) [inline enumerables.rb]: - [ ] Only `each`, `each_with_index`, `select`, and `map` should return `to_enum` when a block is not passed. `all?`, `any?` `none?`, they all do verifications when they are called from an enumerable without a block.

**Use this page as a [guide](https://apidock.com/ruby/Enumerable/any%3F)**

- [ ] When no block or argument is given returns true if at least one of the collection is not false or nil

**expected output:**
Screenshot from 2020-06-10 10-59-12


- [ ] When no block or argument is given returns false if at least one of the collection is not true

**expected output:**
Screenshot from 2020-06-10 11-00-34


- [ ] When a class is passed as an argument returns true if at least one of the collection is a member of such class.

**expected output:**
Screenshot from 2020-06-10 11-03-47

- [ ] When a Regex is passed as an argument returns true if any of the collection matches the Regex

**expected output:**
Screenshot from 2020-06-10 11-05-15

- [ ] When a Regex is passed as an argument returns false if none of the collection matches the Regex

**expected output:**
Screenshot from 2020-06-10 11-06-33

- [ ] When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

**expected output:**
Screenshot from 2020-06-10 11-08-43

- [ ] when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

**expected output:**
Screenshot from 2020-06-10 11-09-59

File: enumerables.rb [hunk 11]
```diff
@@ -1,41 +1,103 @@
...
 
+  def my_none?
+    return to_enum(:my_none?) unless block_given?
+
+    my_each do |x|
+      return false if yield(x) == true
+    end
+    true
+  end
```
Comment (Torres-ssf) [inline enumerables.rb]: - [ ] Only `each`, `each_with_index`, `select`, and `map` should return `to_enum` when a block is not passed. `all?`, `any?` `none?`, they all do verifications when they are called from an enumerable without a block.

**Use this page as a [guide](https://apidock.com/ruby/Enumerable/none%3F)**

- [ ] When no block or argument is given returns true only if none of the collection members is true

**expected output:**
Screenshot from 2020-06-10 11-13-06

- [ ] When no block or argument is given returns false only if one of the collection members is true

**expected output:**
Screenshot from 2020-06-10 11-14-13

- [ ] When a class is passed as an argument returns true if none of the collection is a member of such class

**expected output:**
Screenshot from 2020-06-10 11-18-07

- [ ] When a Regex is passed as an argument returns true if none of the collection matches the Regex

**expected output:**
Screenshot from 2020-06-10 11-22-32

- [ ] When a Regex is passed as an argument returns false if any of the collection matches the Regex

**expected output:**
Screenshot from 2020-06-10 11-24-13

- [ ] When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

**expected output:**
Screenshot from 2020-06-10 11-25-35

- [ ] When a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

**expected output:**
Screenshot from 2020-06-10 11-26-47

File: enumerables.rb [hunk 12]
```diff
@@ -1,41 +1,103 @@
...
+
+  def my_map
+    return to_enum(:my_map) unless block_given?
+
+    result = []
+    my_each do |n|
+      result.push(yield(n))
+    end
+    result
+  end
```
Comment (Torres-ssf) [inline enumerables.rb]: - [ ] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

**expected behavior:**

```
my_proc = proc { |num| num * 5 } 
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10].my_map(&my_proc)
[5, 10, 15, 20, 25, 30, 35, 40, 45, 50]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10].my_map(&my_proc) { |num| num * 2 }
```
The last line should return `[5, 10, 15, 20, 25, 30, 35, 40, 45, 50]` which is the result from proc and not from the block passed as an argument. The block should be ignored when both are passed together.

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Changes Required :muscle: 
Hi there @haraphat01 :hand: 
Great job on this project :+1: 

However, we need to add more functionalities to your methods. We need to do that in order to make them work very similar to the original ones. I tried to add context to each method so you can address all expected behaviors. You can check the documentation [here](https://ruby-doc.org/core-2.7.1/Enumerable.html).

- [ ] Check my comments below.

Kindly submit for code review after making these changes.
Happy coding :computer: 

**Sergio Torres**

**Comment by haraphat01** [comment]
Thank you for the review, here's my final project on enumerables for your kind approval.

Fixed the issue recommended above.

**Review by mikenath223 — CHANGES_REQUESTED** [review]
### Hi @haraphat01  👋

Good job working with previous reviews.
However, let's fix these:

- [ ]  You failed to set up code linting for your repo sorry this wasn't mentioned in previous reviews. Please follow [this link](https://github.com/microverseinc/linters-config/tree/master/ruby) to set up GitHub actions in your repo.

- `my_inject` throws an error of `undefined method: oop?` when any example is passed to it.

- `my_map` should execute **only the proc** passed when both a block and a proc are passed to it.

- `my_none` throws an error of  `undefined method: value` when an Array of entries is passed to it.

- ` my_any`  throws an error of `undefined method: any_checker` when an Array of values is passed to it.

- `my_all` also throws an `undefined method: value` error when an example Array is also passed to it.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(@michgolden) if you need further clarification.

HAPPY CODING👨‍💻

**Comment by mikenath223** [comment]
In addition, kindly resolve the branch conflicts.

**Comment by haraphat01** [comment]
Here is my submission again for the enumerables project for your approval.

I reworked on the;

- My_inject

- my_map

- my_all

- My_none and any and, I also added the Github workflow files.

File: enumerables.rb [hunk 13]
```diff
@@ -1,41 +1,151 @@
...
+      new_object.my_each do |i|
+        new_arr.push(proc.call(i))
+      end
+    else
+      new_object.my_each do |i|
+        new_arr.push(yield(i))
+      end
+    end
+    new_arr
+  end
```
Comment (euqueme) [inline enumerables.rb]: * [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so let's work on it as it was requested by previous reviewers.

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

File: enumerables.rb [hunk 14]
```diff
@@ -1,41 +1,151 @@
...
+        t += 1
+        break if t == new_object.length
+      end
 
+    elsif !args1.nil? && args2.nil?
+      sum = args1
+      new_object.my_each { |i| sum = yield(sum, i) } if block_given?
+    end
+    sum
+  end
```
Comment (euqueme) [inline enumerables.rb]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- raises a "LocalJumpError" when no block or argument is given
```ruby
[1, 5, 8, 10].my_inject == [1, 5, 8, 10].inject 
```
Yours (incorrect)
image

Real inject (correct)
image

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @haraphat01 :lollipop: 

### Thank you for doing the changes requested by the previous reviewers :medal_sports: your project still needs a couple of changes, to be approved so please address to the comments I've left in your code :point_down: 

#### General

- [ ] Activate your project on stickler and add the .rubocop.yml file don't submit for a re-review until your last commit has a green checkmark like this :heavy_check_mark: ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/linter_config.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/54883773/))

### That'ts it :raised_hands: 

### Please :pray: do the changes :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

**Comment by haraphat01** [comment]
Here's my submission again for this project.
I have to revisit the my_inject and map method and also set up the Strickler on the repo.

Thanks

**Comment by lekegitrepo** [comment]
#### Hi @haraphat01 :clap: :clap:
#### You did a good job by implementing the required changes. This project is approved.

#### _Reviewed by Temitayo Adeleke._

**Review by lekegitrepo — APPROVED** [review]
#### Status: Approved.

**Pull request merged**